addons/resource/models/resource.py

```diff
@@ -987,6 +986,129 @@ def _get_unavailable_intervals(self, start, end):
+        personal_leaves_intervals = self._personal_leaves_intervals(start_dt, end_dt)
+        global_leave_per_calendar = self._global_leaves_per_calendar(start_dt, end_dt, calendars_set)
+        dict_attendance_per_calendar = {
+            calendar: calendar._attendance_intervals_batch(start_dt, end_dt, resources=self)
+            for calendar in self.calendar_id
+        }
+        # since the astimezone is a costly operation, we stock the result in a dict in order to avoid the computation in case multiple resources have the same timezone
+        tz_dates = {}
+        for res in self:
+            tz = timezone(res.user_id.tz or res.calendar_id.tz)
+            if (tz, start_dt) in tz_dates:
+                start = tz_dates[(tz, start_dt)]
+            else:
+                start = start_dt.astimezone(tz)
+                tz_dates[(tz, start_dt)] = start
+            if (tz, end_dt) in tz_dates:
+                end = tz_dates[(tz, end_dt)]
+            else:
+                end = end_dt.astimezone(tz)
+                tz_dates[(tz, end_dt)] = end
+
+            personal_leaves = personal_leaves_intervals[res.id] if personal_leaves_intervals.get(res.id) else False
+            global_leaves = global_leave_per_calendar[res.calendar_id.id] if global_leave_per_calendar.get(res.calendar_id.id) else False
+            resource_work_intervals = res._leaves_interval_gantt_view(start, end, personal_leaves, global_leaves, dict_attendance_per_calendar[res.calendar_id][res.id])
+
+            work_intervals = [(start, stop) for start, stop, meta in resource_work_intervals]
+            # start + flatten(intervals) + end
+            work_intervals = [start_datetime] + list(chain.from_iterable(work_intervals)) + [end_datetime]
+            # put it back to UTC
+            work_intervals = list(map(lambda dt: dt.astimezone(utc), work_intervals))
+            # pick groups of two
+            work_intervals = list(zip(work_intervals[0::2], work_intervals[1::2]))
+            leaves_mapping[res.id] = work_intervals
+        return leaves_mapping
+
+    def _personal_leaves_intervals(self, start_dt, end_dt):
+        """
+        :return: a dict { res.id : list[date_start, date_stop] } with the personal leaves of each resource
+        """
+        domain = [
+            ('time_type', '=', 'leave'),
+            ('resource_id', 'in', self.ids),
+            ('date_from', '<=', end_dt),
+            ('date_to', '>=', start_dt),
+        ]
+        leaves = self.env['resource.calendar.leaves'].search_read(domain, fields=['date_from', 'date_to', 'resource_id'])
+        dict_personal_leaves = {}
+        for leave in leaves:
+            if dict_personal_leaves.get(leave['resource_id'][0]):
+                dict_personal_leaves[leave['resource_id'][0]].append([leave['date_from'], leave['date_to']])
+            else:
+                dict_personal_leaves[leave['resource_id'][0]] = [[leave['date_from'], leave['date_to']]]
```

Comment:
```suggestion
        leaves_per_resource = defaultdict(list)
        for leave in leaves:
            dict_personal_leaves[leave['resource_id'][0]].append((leave['date_from'], leave['date_to']))
```
- The default kinda works the same way as `get` (see above). Its arg is the default/"or_else" value it will set on the key that you try to change but isn't yet in the dict.
- Make tuples if you don't plan to update them. :ok_hand: 
- Also, name dict \<value\>_per\<key\> please.